src/dialects/abstract/data-types.ts

```diff
@@ -0,0 +1,1826 @@
+ */
+export type UUIDV4 = _UUIDV4;
+export const UUIDV4 = classToInvokable(_UUIDV4);
+
+class _VIRTUAL<T> extends ABSTRACT<T> {
+  public static readonly key = 'VIRTUAL';
+
+  returnType?: ABSTRACT<T>;
+  fields?: string[];
+
+  /**
+   * @param [ReturnType] return type for virtual type
+   * @param [fields] array of fields this virtual type is dependent on
+   */
+  constructor(ReturnType?: DataType, fields?: string[]) {
+    super();
+    if (typeof ReturnType === 'function') {
+      ReturnType = new ReturnType();
+    }
+
+    this.returnType = ReturnType;
+    this.fields = fields;
+  }
+}
+
+/**
+ * A virtual value that is not stored in the DB. This could for example be useful if you want to provide a default value in your model that is returned to the user but not stored in the DB.
+ *
+ * You could also use it to validate a value before permuting and storing it. VIRTUAL also takes a return type and dependency fields as arguments
+ * If a virtual attribute is present in `attributes` it will automatically pull in the extra fields as well.
+ * Return type is mostly useful for setups that rely on types like GraphQL.
+ *
+ * @example <caption>Checking password length before hashing it</caption>
+ * sequelize.define('user', {
+ *   password_hash: DataTypes.STRING,
+ *   password: {
+ *     type: DataTypes.VIRTUAL,
+ *     set: function (val) {
+ *        // Remember to set the data value, otherwise it won't be validated
+ *        this.setDataValue('password', val);
+ *        this.setDataValue('password_hash', this.salt + val);
+ *      },
+ *      validate: {
+ *         isLongEnough: function (val) {
+ *           if (val.length < 7) {
+ *             throw new Error("Please choose a longer password")
+ *          }
+ *       }
+ *     }
+ *   }
+ * })
+ *
+ * # In the above code the password is stored plainly in the password field so it can be validated, but is never stored in the DB.
+ *
+ * @example <caption>Virtual with dependency fields</caption>
+ * {
+ *   active: {
+ *     type: new DataTypes.VIRTUAL(DataTypes.BOOLEAN, ['createdAt']),
+ *     get: function() {
+ *       return this.get('createdAt') > Date.now() - (7 * 24 * 60 * 60 * 1000)
+ *     }
+ *   }
+ * }
+ *
+ */
+export type VIRTUAL<T> = _VIRTUAL<T>;
+export const VIRTUAL = classToInvokable(_VIRTUAL);
+
+interface EnumOptions<Member extends string> {
+  values: Member[];
+}
+
+type EnumParams<Member extends string> =
+  | []
+  | [EnumOptions<Member>]
+  | [Member[]]
+  | Array<Member | Member[]>;
+
+class _ENUM<Member extends string> extends ABSTRACT<Member> {
+  public static readonly key = 'ENUM';
+  public values: Member[];
+  protected options: EnumOptions<Member>;
+
+  /**
+   * @param {...any|{ values: any[] }|any[]} args either array of values or options object with values array. It also supports variadic values
+   */
+  constructor(options: EnumOptions<Member>);
+  constructor(members: Member[]);
+  constructor(...members: Array<Member | Member[]>);
```

Comment:
`DataTypes.ENUM('a', ['b', 'c'])` shouldn't be accepted

```suggestion
  constructor(...members: Member[]);
```